### PR TITLE
feat(core): restrict global and table imports in program code

### LINKED ISSUE #3736: Make instantiation error impossible in core-processor
### Problem to Solve

Currently we have two places gear_core::code and core_backend::env where our code is checked to be suitable for execution. We can make almost all checks in `Code::try_new`.

Unfortunately, we cannot guaranty that `Instance::new` would return always `Ok(())`, because executors have it's own validators.

### Possible Solution

Currently should be added:
1) excess imports
2) checks for stack end, see https://github.com/gear-tech/gear/issues/3735

Also, we can use Environment::new() for code checking, but the result can be different for different executors and currently we cannot use wasmer as embedded executor in tests for example.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by ByteNacked**
Resolves #3736.

Add additional import checks to `check_imports`.

Added checks:
* There are no global or table imports in the module.

@reviewer-or-team

### CODE REVIEW

File: core/src/code/errors.rs [hunk 1]
```diff
@@ -132,6 +132,20 @@ pub enum ImportError {
     /// The signature of an imported function is invalid.
     #[display(fmt = "Invalid function signature for imported function with index `{_0}`")]
     InvalidImportFnSignature(u32),
+    /// Unexpected import kind.
+    #[display(fmt = "Unexpected import kind `{kind}` with index `{index}`")]
+    UnexpectedImportKind {
+        /// Kind of the import.
+        kind: &'static str,
```
Comment (StackOverflowExcept1on) [inline core/src/code/errors.rs:139]: `&'static str` is fat pointer, plz use `&'static &'static str` or enum

Comment (ByteNacked) [reply core/src/code/errors.rs]: It looks like &str is better option here

Fixed

File: core/src/code/utils.rs [hunk 2]
```diff
@@ -160,37 +160,64 @@ pub fn check_imports(module: &Module) -> Result<(), CodeError> {
...
-            .ok_or(ImportError::UnknownImport(import_index as u32))?;
 
-        if !visited_imports.insert(*syscall) {
-            Err(ImportError::DuplicateImport(import_index as u32))?;
+    for (import_index, import) in imports.iter().enumerate() {
+        let import_index: u32 = import_index
+            .try_into()
+            .unwrap_or_else(|_| unreachable!("Import index should fit in u32"));
+
+        match import.external() {
```
Comment (StackOverflowExcept1on) [inline core/src/code/utils.rs:172]: - add new tests to `core/src/code/mod.rs` for cases when a global/table/2 memories is imported
- remove https://github.com/gear-tech/gear/blob/8ba3c6996be2c4fd060f2d9f16b292c72b201f2c/pallets/gear/src/tests.rs#L13055-L13057

**Review by StackOverflowExcept1on — APPROVED** [review]

**Comment by ByteNacked** [comment]
Waiting for 1 more reviewer's approval

File: core/src/code/utils.rs [hunk 3]
```diff
@@ -161,35 +161,53 @@ pub fn check_imports(module: &Module) -> Result<(), CodeError> {
...
+            kind @ (External::Global(_) | External::Table(_)) => {
+                Err(ImportError::UnexpectedImportKind {
+                    kind: if matches!(kind, External::Global(_)) {
+                        &"Global"
+                    } else {
+                        &"Table"
+                    },
+                    index: import_index,
+                })?
+            }
```
Comment (grishasobol) [inline core/src/code/utils.rs:209]: This place looks a little tricky. Change or not - by your choice
```suggestion
            External::Global(_) => Err(ImportError::UnexpectedImportKind {
                kind: "Global",
                index: import_index,
            })?,
            External::Table(_) => Err(ImportError::UnexpectedImportKind {
                kind: "Table",
                index: import_index,
            })?,
```

File: core/src/code/errors.rs [hunk 4]
```diff
@@ -132,6 +132,14 @@ pub enum ImportError {
...
     #[display(fmt = "Invalid function signature for imported function with index `{_0}`")]
     InvalidImportFnSignature(u32),
+    /// Unexpected import kind.
+    #[display(fmt = "Unexpected import kind `{kind}` with index `{index}`")]
+    UnexpectedImportKind {
+        /// Kind of the import.
+        kind: &'static &'static str,
+        /// Index of the import.
+        index: u32,
+    },
```
Comment (grishasobol) [inline core/src/code/errors.rs:142]: I cannot understand why `kind` has type `&'static &'static str`. If it's for optimisation purposes, than `&'static str` is also pointer and have the same size.

File: core/src/code/utils.rs (hunk 3 above)
Comment (ByteNacked) [reply core/src/code/utils.rs:209]: I personally prefer a simpler and more boilerplate variant, so I'm going to use your suggestion.

File: core/src/code/errors.rs (hunk 4 above)
Comment (ByteNacked) [reply core/src/code/errors.rs:142]: https://github.com/gear-tech/gear/pull/3911#discussion_r1574580698

They'are not if I understand you correctly

Comment (ByteNacked) [reply core/src/code/errors.rs:142]: &'static str is a fat ptr, &'static &'static str is plain ptr

Comment (ByteNacked) [reply core/src/code/errors.rs:142]: Not sure if it's worth it for optimization purposes though.

File: core/src/code/utils.rs (hunk 3 above)
Comment (ByteNacked) [reply core/src/code/utils.rs]: Fixed

File: core/src/code/errors.rs (hunk 4 above)
Comment (grishasobol) [reply core/src/code/errors.rs:142]: Yes, forgot about this. Ok

**Review by grishasobol — APPROVED** [review]

**Comment by breathx** [comment]
👍🏻

**Pull request merged**